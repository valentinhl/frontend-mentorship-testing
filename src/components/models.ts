export type FormValueListType = {
  [key: string]: number | string | undefined | null | File;
};

export type ListItemType = FormValueListType & {
  id: string;
  title: string;
  description?: string;
};
export type ListFormOnSubmit = (newValue: ListItemType) => void;
export type ListFormOnChange = (newFormItem: ListItemType) => void;
export interface ListItemFormData {
  title: { value: string };
  description: { value: string };
}
export interface ListItemFormErrorList {
  title: string;
  description: string;
}

export type CreateEmptyErrorList = () => ListItemFormErrorList;
