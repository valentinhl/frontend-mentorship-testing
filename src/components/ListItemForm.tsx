import { FC } from 'react';
import { ListItemType } from './models';
import { v4 as uuidv4 } from 'uuid';
import {
  descriptionLabel,
  submitButtonText,
  titleLabel
} from './ListItemUIText';
import './ListItemForm.css';
import useForm from '../hooks/useForm';

export type ListFormOnSubmit = (newValue: ListItemType) => void;

export interface ListItemFormProps {
  onSubmit: ListFormOnSubmit;
  isSubmitDisabled?: boolean;
}

export interface ListItemFormData {
  title: { value: string };
  description: { value: string };
}

const ListItemForm: FC<ListItemFormProps> = ({
  onSubmit,
  isSubmitDisabled = false
}): JSX.Element => {
  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const { title, description } = (
      event.target as EventTarget & HTMLFormElement
    ).elements as HTMLFormControlsCollection & ListItemFormData;

    onSubmit({
      id: uuidv4(),
      title: title.value,
      description: description.value
    });
  };

  const { handleChange, values, errors } = useForm({ callback: submitForm });

  return (
    <form className='todo-creation-form' onSubmit={event => submitForm(event)}>
      <fieldset className='fieldset'>
        <label htmlFor='title'>{titleLabel}</label>
        <input
          className='input-field'
          id='title'
          name='title'
          onChange={handleChange}
        />
        {errors.title && <span className='error'>{errors.title}</span>}
      </fieldset>

      <fieldset className='fieldset'>
        <label htmlFor='description'>{descriptionLabel}</label>
        <textarea
          className='input-field'
          id='description'
          placeholder='Please enter a description here'
          name='description'
          onChange={handleChange}
        />
        {errors.description && (
          <span className='error'>{errors.description}</span>
        )}
      </fieldset>

      <button disabled={isSubmitDisabled}>{submitButtonText}</button>
    </form>
  );
};

export default ListItemForm;
