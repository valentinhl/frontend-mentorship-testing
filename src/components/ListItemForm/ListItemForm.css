.todo-creation-form {
  display: flex;
  flex-direction: column;
  background-color: rgba(255, 255, 255, 0.75);
  border-radius: 32px;
  padding: 32px;
  margin-top: 32px;
  min-width: 350px;
}

.fieldset {
  border: none;
  display: flex;
  justify-content: space-between;
}

.input-field {
  width: 200px;
}

.error {
  display: 'block';
  color: red;
}
