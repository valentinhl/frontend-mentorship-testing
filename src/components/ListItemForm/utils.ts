import { CreateEmptyErrorList, ListItemFormErrorList } from '../models';

export const isRequired = (
  errors: ListItemFormErrorList,
  field: 'title' | 'description',
  value: string
): ListItemFormErrorList => {
  const allowedFields = ['title', 'description'];
  const newErrors = { ...errors };
  if (allowedFields.includes(field)) {
    const fieldError = !value.length ? `${field} must be entered` : '';
    newErrors[field] = fieldError;
  }
  return newErrors;
};

export const createEmptyErrorList: CreateEmptyErrorList = () => ({
  title: '',
  description: ''
});
