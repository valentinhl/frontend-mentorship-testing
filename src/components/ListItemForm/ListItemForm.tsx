import { FC, useState } from 'react';
import {
  ListFormOnChange,
  ListFormOnSubmit,
  ListItemFormData,
  ListItemFormErrorList,
  ListItemType
} from '../models';
import { v4 as uuidv4 } from 'uuid';
import {
  addListItemForm,
  descriptionLabel,
  submitButtonText,
  titleLabel
} from '../ListItemUIText';
import './ListItemForm.css';
import { createEmptyErrorList, isRequired } from './utils';
export interface ListItemFormProps {
  formItem: ListItemType;
  onChange: ListFormOnChange;
  onSubmit: ListFormOnSubmit;
  isSubmitDisabled?: boolean;
}

const ListItemForm: FC<ListItemFormProps> = ({
  formItem,
  onChange,
  onSubmit,
  isSubmitDisabled = false
}): JSX.Element => {
  const [errors, setErrors] = useState<ListItemFormErrorList>(
    createEmptyErrorList()
  );

  const validate = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    name: string,
    value: string
  ) => {
    setErrors(isRequired(errors, name as 'title' | 'description', value));
  };

  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    event.persist();

    const name = event.target.name;
    const val = event.target.value;

    validate(event, name, val);

    const newValue = { ...formItem };
    newValue[name] = val;

    onChange(newValue);
  };

  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const { title, description } = (
      event.target as EventTarget & HTMLFormElement
    ).elements as HTMLFormControlsCollection & ListItemFormData;

    onSubmit({
      id: formItem.id ? formItem.id : uuidv4(),
      title: title.value,
      description: description.value
    });
  };

  return (
    <form
      name={addListItemForm}
      className='todo-creation-form'
      onSubmit={event => submitForm(event)}
    >
      <fieldset className='fieldset'>
        <label htmlFor='title'>{titleLabel}</label>
        <input
          className='input-field'
          id='title'
          name='title'
          onChange={handleChange}
          value={formItem.title}
        />
        {errors.title && <span className='error'>{errors.title}</span>}
      </fieldset>

      <fieldset className='fieldset'>
        <label htmlFor='description'>{descriptionLabel}</label>
        <textarea
          className='input-field'
          id='description'
          placeholder='Please enter a description here'
          name='description'
          onChange={handleChange}
          value={formItem.description}
        />
        {errors.description && (
          <span className='error'>{errors.description}</span>
        )}
      </fieldset>

      <button disabled={isSubmitDisabled}>{submitButtonText}</button>
    </form>
  );
};

export default ListItemForm;
