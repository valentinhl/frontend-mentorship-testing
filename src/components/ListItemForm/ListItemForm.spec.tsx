import { render, screen, cleanup, RenderResult } from '@testing-library/react';
import ListItemForm from './ListItemForm';
import { MOCK_STRING } from '../../testing/mockValues';
import { typeInputValue, clickButton } from '../../testing/utils';
import {
  submitButtonText,
  titleLabel
} from '../ListItemUIText';
import { ListItemType, ListFormOnChange, ListFormOnSubmit } from '../models';

const getMockId = () => 'id';

jest.mock('uuid', () => ({
  ...jest.requireActual('uuid'),
  v4: getMockId
}));

describe('ListItemForm.tsx', () => {
  const MOCK_ID = getMockId();
  const onSubmitMock = jest.fn() as jest.MockedFunction<ListFormOnSubmit>;
  const onChangeMock = jest.fn() as jest.MockedFunction<ListFormOnChange>;
  const formItemMock: ListItemType = {
    id: MOCK_ID,
    title: '',
    description: ''
  };
  let rerenderComponent: RenderResult['rerender'];

  beforeEach(() => {
    const { rerender } = render(
      <ListItemForm
        onSubmit={onSubmitMock}
        onChange={onChangeMock}
        formItem={formItemMock}
      />
    );
    rerenderComponent = rerender;
  });

  afterEach(() => {
    onSubmitMock.mockClear();
    cleanup();
  });

  it('renders the component successfully', () => {
    expect(screen.getByRole('form')).toMatchSnapshot();
  });

  it('submit is enabled by default', () => {
    expect(screen.getByText(submitButtonText)).toBeEnabled();
  });

  it('submit is enabled when isSubmitDisabled is passed', () => {
    rerenderComponent(
      <ListItemForm
        onSubmit={onSubmitMock}
        onChange={onChangeMock}
        formItem={formItemMock}
        isSubmitDisabled
      />
    );
    expect(screen.getByText(submitButtonText)).toBeDisabled();
  });

  it('submit is enabled when isSubmitDisabled is false', () => {
    rerenderComponent(
      <ListItemForm
        onSubmit={onSubmitMock}
        onChange={onChangeMock}
        formItem={formItemMock}
        isSubmitDisabled={false}
      />
    );
    expect(screen.getByText(submitButtonText)).toBeEnabled();
  });

  it('calls onChange prop when change event fires', () => {
    typeInputValue(screen, titleLabel, MOCK_STRING);
    expect(onChangeMock).toHaveBeenCalled();
  });

  it('calls onSubmit prop once when submit event fires', () => {
    clickButton(screen, submitButtonText);
    expect(onSubmitMock).toHaveBeenCalledTimes(1);
  });

  it('onSubmit receives object with right values', () => {
    rerenderComponent(
      <ListItemForm
        onSubmit={onSubmitMock}
        onChange={onChangeMock}
        formItem={{
          id: MOCK_ID,
          title: MOCK_STRING,
          description: MOCK_STRING
        }}
      />
    );

    clickButton(screen, submitButtonText);

    expect(onSubmitMock).toHaveBeenCalledWith({
      id: MOCK_ID,
      title: MOCK_STRING,
      description: MOCK_STRING
    });
  });
});
