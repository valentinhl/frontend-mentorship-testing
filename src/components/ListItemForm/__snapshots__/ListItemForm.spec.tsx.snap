// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemForm.tsx renders the component successfully 1`] = `
<form
  class="todo-creation-form"
  name="Add a new list item form"
>
  <fieldset
    class="fieldset"
  >
    <label
      for="title"
    >
      Title:
    </label>
    <input
      class="input-field"
      id="title"
      name="title"
      value="MOCK_STRING"
    />
    
  </fieldset>
  <fieldset
    class="fieldset"
  >
    <label
      for="description"
    >
      Description:
    </label>
    <textarea
      class="input-field"
      id="description"
      name="description"
      placeholder="Please enter a description here"
    >
      MOCK_STRING
    </textarea>
    
  </fieldset>
  <button>
    SUBMIT
  </button>
</form>
`;
